[SPARK-17159] [streaming]: optimise check for new files in FileInputDStream

steveloughran opened the pull request:
## What changes were proposed in this pull request?

This PR optimises the filesystem metadata reads in `FileInputDStream`, by moving the filters used in `FileSystem.globStatus` and `FileSystem.listStatus` into filtering of the `FileStatus` instances returned in the results, so avoiding the need to create `FileStatus` instances within the `FileSystem` operation.
- This doesn't add overhead to the filtering process; that's done as post-processing in the`FileSystem` glob/list operations anyway.
- At worst it may result in larger lists being built up and returned.
- For every glob match of a file, the code saves 1 RPC calls to the HDFS NN; 1 GET against S3
- For every glob match of a directory, the code the code saves 1 RPC call and 2-3 HTTP calls to S3 for the directory check (including a slow List call whenever the directory has children as it doesn't exist as a blob any more)
- for the modtime check of every file, it saves a Hadoop RPC call, against all object stores _which don't implement any client-side cache_, an HTTP GET. 
- By entirely eliminating all `getFileStatus()` calls in the listed files, it should reduce the risk of AWS S3 throttling the HTTP request, as it does when too many requests are made to parts of a single S3 bucket.
## How was this patch tested?

Running the spark streaming tests as a regression suite. In the SPARK-7481 cloud code, I could add a test against S3 which prints to stdout the exact number of HTTP requests made to S3 before and after the patch, so as to validate speedup. (the S3A metrics in Hadoop 2.8+ are accessible at the API level, but as they are only accessible in a new API added in 2.8; it'd stop that proposed module building against Hadoop 2.7. Logging and manual assessment is the only cross-version strategy.

steveloughran committed:
SPARK-17159: move filtering of directories and files out of glob/list filters and into filtering of the FileStatus instances returned in the results, so avoiding the need to create FileStatus intances for

-This doesn't add overhead to the filtering process; that's done as post-processing in FileSystem anyway. At worst it may result in larger lists being built up and returned.
-For every glob match, the code saves 2 RPC calls to the HDFS NN
-The code saves 1-3 HTTP calls to S3 for the directory check (including a slow List call whenever the directory has children as it doesn't exist as a blob any more)
-for the modtime check of every file, it saves an HTTP GET

The whole modtime cache can be eliminated; it's a performance optimisation to avoid the overhead of the file checks, one that is no longer needed.

SparkQA commented:
**[Test build #64140 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/64140/consoleFull)** for PR 14731 at commit [`738c51b`](https://github.com/apache/spark/commit/738c51bb57f331c58a877aa20aa5e2beb1084114).

SparkQA commented:
**[Test build #64140 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/64140/consoleFull)** for PR 14731 at commit [`738c51b`](https://github.com/apache/spark/commit/738c51bb57f331c58a877aa20aa5e2beb1084114).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/64140/
Test PASSed.

srowen commented:
LGTM. Does this sort of change make sense elsewhere where `PathFilter` is used? I glanced at the others and it looked like a wash in other cases.

steveloughran commented:
# I'm going to scan through and tune them elsewhere; really I'm going by uses of the listFiles calls

There's actually no significant use elsewhere that I can see; just a couple of uses which filter on filename —so there is no cost penalty.

* `SparkHadoopUtil.listLeafStatuses()` does implement its own directory recursion to find files; FileSystem.listFiles(path, true) does that, and on S3A will do flat scan that is O(files/5000); no directory overhead at all.
* Otherwise, globStatus() can be pretty slow against object stores, but the fix there isn't in the client code; it means someone needs to implement [HADOOP-13371](https://issues.apache.org/jira/browse/HADOOP-13371), *S3A globber to use bulk listObject call over recursive directory scan* —more specifically, an implementation scalable to production datasets. 

Returning to this patch, should I cut out the caching? I think it is superfluous.

srowen commented:
Why is the caching superfluous -- because no file is evaluated more than once here?

steveloughran commented:
to be precise: the caching of file modification times is superfluous. It's there to avoid the cost of executing `getFileStatus()` on previously scanned files. Once you use the FileStatus returned in a listing, you aren't calling `getFileStatus()`, hence: no need to cache

steveloughran committed:
[SPARK-17159] Remove the fileModTime cache. Now that the modification time costs 0 to evaluate, caching it actually consumes memory and the time for a lookup.

SparkQA commented:
**[Test build #64142 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/64142/consoleFull)** for PR 14731 at commit [`6e8ace0`](https://github.com/apache/spark/commit/6e8ace0444ec9bdebc7c809a08628891f6de5fd0).

srowen commented:
Ah right, you already have the modification time for free. Sounds good, remove the caching.

SparkQA commented:
**[Test build #64142 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/64142/consoleFull)** for PR 14731 at commit [`6e8ace0`](https://github.com/apache/spark/commit/6e8ace0444ec9bdebc7c809a08628891f6de5fd0).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/64142/
Test PASSed.

petermaxlee commented on streaming/src/main/scala/org/apache/spark/streaming/dstream/FileInputDStream.scala:
```diff
@@ -241,16 +233,21 @@ class FileInputDStream[K, V, F <: NewInputFormat[K, V]](
+   * @return true if the file has been modified within the batch window
    */
-  private def isNewFile(path: Path, currentTime: Long, modTimeIgnoreThreshold: Long): Boolean = {
+ private def isNewFile(fs: FileStatus, currentTime: Long, modTimeIgnoreThreshold: Long): Boolean = {
```
indent is wrong here

petermaxlee replied:
also fs is pretty confusing, because in this context it is often used to refer to as FileSystem. We should pick a different word.

steveloughran replied:
I'll fix this

petermaxlee commented on streaming/src/main/scala/org/apache/spark/streaming/dstream/FileInputDStream.scala:
```diff
@@ -293,8 +290,8 @@ class FileInputDStream[K, V, F <: NewInputFormat[K, V]](
   /** Get file mod time from cache or fetch it from the file system */
-  private def getFileModTime(path: Path) = {
-    fileToModTime.getOrElseUpdate(path.toString, fs.getFileStatus(path).getModificationTime())
+  private def getFileModTime(fs: FileStatus) = {
```
should we just remove this function now?

steveloughran replied:
yes, I was just being minimal about the changes. Inlining is easy

steveloughran committed:
[SPARK-17159] inline FileStatus.getModificationTime; address style issues. Also note that 1s granularity is the resolution from HDFS; other filesystems may have a different resolution. The only one I know that is worse is FAT16/FAT32, which is accurate to 2s, but nobody should be using that except on SSD cards and USB sticks

SparkQA commented:
**[Test build #64156 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/64156/consoleFull)** for PR 14731 at commit [`b08e3c9`](https://github.com/apache/spark/commit/b08e3c9937a63a08b274a1491ea7064168646f1d).

SparkQA commented:
**[Test build #64156 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/64156/consoleFull)** for PR 14731 at commit [`b08e3c9`](https://github.com/apache/spark/commit/b08e3c9937a63a08b274a1491ea7064168646f1d).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/64156/
Test PASSed.

srowen commented:
This is ready to go right @steveloughran ? LGTM

steveloughran commented:
LGTM. I was trying to see if there was a way to create a good test here by triggering the takes-too-long codepath and having a counter, but there's no obvious way to do that deterministically. I am doing a test for this against s3 in the spark-cloud module I'm writing; I can look at the printed counts of getFileStatus before/after the patch to see the difference, but the actual (testable) metrics are only accessible with forthcoming Hadoop 2.8 release.

TL;DR: no easy test, so there's nothing left to do

steveloughran commented:
Actually, I've just noticed that DStream behaviour isn't in sync with the streaming programming guide, which says "files written in nested directories not supported)". That is: SPARK-14796 didn't patch the docs.

it may as well be fixed in this patch. How about, in the bullet points underneath

- Wildcards may be used to specify a set of directories to scan for new files, for example `hdfs://nn1:8050/users/alice/logs/2016-*/*.gz`
-
-New directories and their contents will be discovered as they arrive

Special points for object stores
- Wildcard lookup may be very slow with some object stores.
 - Directory rename is not atomic; if a directory is renamed into the streaming source, then the files within may only be discovered and process across a multiple streaming windows.
- 

+ there's another optimisation; use the {{SparkHadoopUtils.isGlobPath()}} predicate to recognise when the dir path isn't a wildcard, in which case just do a simple listFiles()}}. Until that shortcutting is done automatically in the Hadoop FS implementation, spark can do it on its side. As the {{listFiles()}} call was what was used before SPARK-14796, it has to be compatible, else SPARK-14796 has broken things

steveloughran commented:
I've now done the [s3a streaming test/example](https://github.com/steveloughran/spark/blob/features/SPARK-7481-cloud/cloud/src/main/scala/org/apache/spark/cloud/s3/examples/S3Streaming.scala)

this uses a pattern of s3a/path/sub* as the directory path; then creates a file in a directory and renames the dir to match the path; verifies that the file was found in the time period allocated

https://gist.github.com/steveloughran/c8b39a7b87a9bd63d7a383bda8687e7e


Notable that the scan of the empty dir took 150ms; once there's data in the tree the time jumps up to 500ms once there are two entries under the tree, one dir and one file

summary stats show 72 getFileStatus calls at the FS API, mapping to 140 HEAD calls and 88 LIST operations. 

```
 S3AFileSystem{uri=s3a://stevel-ireland-new, workingDir=s3a://hwdev-steve-ireland-new/user/stevel, inputPolicy=sequential, partSize=104857600, enableMultiObjectsDelete=true, maxKeys=5000, readAhead=65536, blockSize=1048576, multiPartThreshold=2147483647, statistics {292 bytes read, 292 bytes written, 101 read ops, 0 large read ops, 11 write ops}, metrics {{Context=S3AFileSystem} {FileSystemId=343b706a-c238-4d71-9ed8-8083601ac28a-hwdev-steve-ireland-new} {fsURI=s3a://hwdev-steve-ireland-new} {files_created=1} {files_copied=1} {files_copied_bytes=292} {files_deleted=1} {directories_created=3} {directories_deleted=0} {ignored_errors=2} {op_copy_from_local_file=0} {op_exists=1} {op_get_file_status=72} {op_glob_status=16} {op_is_directory=0} {op_is_file=0} {op_list_files=0} {op_list_located_status=0} {op_list_status=27} {op_mkdirs=2} {op_rename=1} {object_copy_requests=0} {object_delete_requests=3} {object_list_requests=88} {object_continue_list_requests=0} {object_metadata_requests=140} {object_multipart_aborted=0} {object_put_bytes=292} {object_put_requests=4} {stream_read_fully_operations=0} {stream_bytes_skipped_on_seek=0} {stream_bytes_backwards_on_seek=0} {stream_bytes_read=292} {streamOpened=1} {stream_backward_seek_pperations=0} {stream_read_operations_incomplete=0} {stream_bytes_discarded_in_abort=0} {stream_close_operations=1} {stream_read_operations=1} {stream_aborted=0} {stream_forward_seek_operations=0} {streamClosed=1} {stream_seek_operations=0} {stream_bytes_read_in_close=0} {stream_read_exceptions=0} }}
```

I'm going to do a test run with the modification here and see what it does to listing and status

steveloughran committed:
[SPARK-17159] updates as discussed on PR: skip wildcards for non wildcarded listing; handle FNFE specially, add the docs

SparkQA commented:
**[Test build #64296 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/64296/consoleFull)** for PR 14731 at commit [`79b57a2`](https://github.com/apache/spark/commit/79b57a2683dece86e1acd063b2d33fa5a6dd6038).

steveloughran commented:
1. updated the code to bypass the glob routine when there is no wildcard; this bypasses something fairly inefficient. 
1. reporting FNFE on that base dir differently; skip the stack trace (maybe: log at a lower level?). 
1. Updated the docs with a special list of blobstore best practises.

It's a bit hard to get some of that phrasing of what the wildcard does right; needs careful review.

Tested using my s3 streaming test, which did use a * in the wildcard. All works, but no improvements in speed on what is a fairly unrealistic structure. The time to recursively list object stores remotely is tangibly slow. Maybe that should go in the text too: "it can be take seconds to scan object stores for new data, with the time being proportional to directory depth and the number of files in a directory. Shallow and wide directory trees are faster"

SparkQA commented:
**[Test build #64296 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/64296/consoleFull)** for PR 14731 at commit [`79b57a2`](https://github.com/apache/spark/commit/79b57a2683dece86e1acd063b2d33fa5a6dd6038).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/64296/
Test PASSed.

srowen commented on docs/streaming-programming-guide.md:
```diff
@@ -644,13 +644,39 @@ methods for creating DStreams from files as input sources.
+     + A simple directory can be supplied (`hdfs://namenode:8040/logs/`). All files directly
+       underneath this path will be processed as they are discovered.
+     + A regular expression can be supplied instead, such as
+       `s3a://bucket/logs/[2015,2016]-??-??-friday`.
```
This isn't a regular expression though, what was the intent here?

steveloughran replied:
wasn't it? My mistake. I wanted to show something fairly complex.

srowen replied:
`(2015|2016)` would be what means "the string '2015' or '2016'", and `..` would mean "any two characters"

steveloughran replied:
going with `s3a://bucket/logs/(2015,2016)-*-friday`

srowen replied:
You'll need (2015|2016) rather than (2015,2016). Also this is going to match zero or more hyphens followed by "-friday". I think you mean ".." or ".{2}" or at least ".+" instead of "*" if this is a regex.

steveloughran replied:
Round () or {} brackets? Because the wildcard pattern used in `isGlobPath()` is `"{}[]*?\\"`. Curly only.

srowen replied:
Oh, is this not a regular expression? I'd change the doc then to not describe it as one. It sounds like it's following some other kind of glob syntax.

srowen commented on streaming/src/main/scala/org/apache/spark/streaming/dstream/FileInputDStream.scala:
```diff
@@ -196,29 +192,33 @@ class FileInputDStream[K, V, F <: NewInputFormat[K, V]](
+            .filter(_.isDirectory)
+            .map(_.getPath)
+      } else {
+        List(directoryPath).toArray
```
Just `Array(directoryPath)`?

steveloughran replied:
yes, but obsolete right now with the refactoring in progress

steveloughran commented:
The logic has got complex enough it merits unit tests. Pulling into SparkHadoopUtils itself and writing some for the possible: simple, glob matches one , glob matches 1+, glob doesn't match, file not found

steveloughran committed:
[SPARK-17159] move glob operation into SparkHadoopUtils, alongside an existing/similar method. Add tests for the behaviour. Update docs with suggested fixes, and review/edit.

SparkQA commented:
**[Test build #64368 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/64368/consoleFull)** for PR 14731 at commit [`b63abfe`](https://github.com/apache/spark/commit/b63abfe32a5509f69c7f725a46b2e6ac8fb9cf1f).

steveloughran commented:
Having looked at the source code, `FileSystem.globStatus()` uses the glob patterns, which are not the same as the posix regexp ones. [org.apache.hadoop.fs.GlobPattern](http://grepcode.com/file/repo1.maven.org/maven2/org.apache.hadoop/hadoop-common/2.7.1/org/apache/hadoop/fs/GlobPattern.java#81) does the conversion.

For the docs, I'll just use a wildcard * in the example, rather than try anything more sophisticated.

steveloughran committed:
[SPARK-17159]  use a simpler pattern in the example docs

SparkQA commented:
**[Test build #64368 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/64368/consoleFull)** for PR 14731 at commit [`b63abfe`](https://github.com/apache/spark/commit/b63abfe32a5509f69c7f725a46b2e6ac8fb9cf1f).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/64368/
Test FAILed.

steveloughran committed:
[SPARK-17159] add directory rename test (taken from SPARK-7481 examples and made more robust)

SparkQA commented:
**[Test build #64486 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/64486/consoleFull)** for PR 14731 at commit [`9bc0ea9`](https://github.com/apache/spark/commit/9bc0ea9734ccaf11c6306a3496c98be8cc20faab).

SparkQA commented:
**[Test build #64486 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/64486/consoleFull)** for PR 14731 at commit [`9bc0ea9`](https://github.com/apache/spark/commit/9bc0ea9734ccaf11c6306a3496c98be8cc20faab).
 * This patch **fails Scala style tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/64486/
Test FAILed.

steveloughran committed:
[SPARK-17159] method nested inside a sparktest test closure being mistaken for a public method and so needing to declare a return type.

SparkQA commented:
**[Test build #64488 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/64488/consoleFull)** for PR 14731 at commit [`fe40bd2`](https://github.com/apache/spark/commit/fe40bd2bf548ca973f9dcdf9426fb9834828f72b).

SparkQA commented:
**[Test build #64488 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/64488/consoleFull)** for PR 14731 at commit [`fe40bd2`](https://github.com/apache/spark/commit/fe40bd2bf548ca973f9dcdf9426fb9834828f72b).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/64488/
Test PASSed.

srowen commented on core/src/main/scala/org/apache/spark/deploy/SparkHadoopUtil.scala:
```diff
@@ -244,6 +244,31 @@ class SparkHadoopUtil extends Logging {
+   * @param pattern pattern
+   * @return a possibly empty array of FileStatus entries
+   */
+  def globToFileStatus(pattern: Path): Array[FileStatus] = {
```
I'm concerned that this method is necessarily ambiguous, because you can't actually distinguish globs from other paths. Is this really needed? that's why the FS API exposes two methods.

steveloughran replied:
It goes with the `globPathIfNecessary` call —you want to rename it to be consistent.


Regarding the FS APIs, There's way to many list operations in the FS APIs, each with different flaws.

1. The simple `list(path, filter): Array[FS]` operations don't scale to a directory with hundreds of thousands of files, hence the remote iterator versions
1. None of them provide any consistency guarantees. Worth knowing. This is more common in remote iterators as the iteration window is bigger, but even in those that return arrays, in a large enough directory things may change during the enum
1. Anything that treewaks is very suboptimal on blobstores, somewhat inefficient for deep trees.
1. `listFiles(path, recursive=true)` is the sole one which object stores can currently optimise by avoiding the treewalk and just doing a bulk list. [HADOOP-13208](https://issues.apache.org/jira/browse/HADOOP-13208) has added that for S3A.
1. ..but that method filters out all directories, which means that apps which do want directories too are out of luck.
1. globStatus() is even less efficient than the others ... have a look at the source to see why.
1. In [HADOOP-13371](https://issues.apache.org/jira/browse/HADOOP-13371) I'm exploring an optimised globber, but I don't want to write one which collapses at scale (i.e in production).

I've added some comments in HADOOP-13371 about what to do there, I will probably do that "no regexp -> simple return" strategy implemented in this patch. But it will only benefit s3a in Hadoop 2.8+; patching spark benefits everything.

srowen replied:
Here I'm narrowly concerned with the ambiguity of the behavior of a single method, because you can't distinguish between a path with a "?" in it and a glob wildcard for example. The rest seems orthogonal?

The change as it stood to resolve the issue in the OP seemed OK. This is bigger now and I'm not as sure about the rest of the change.

steveloughran replied:
essentially if anything which might be a wildcard is hit, it gets handed off to the globber for the full interpretation. Same for ^ and ], which are only part of a pattern within the context of an opening [

Its only those strings which can be verified to be regexp free in a simple context-free string scan that say "absolutely no patterns here"

regarding the bigger change: most of it is isolation of the sensitive code *and the tests to verify behaviour*

srowen replied:
Yea, but then that's wrong if for example my path actually has a ? or ^ or ] in it. It doesn't seem essential and seems even problematic to add this behavior change to an otherwise clear fix.

steveloughran replied:
I recognise your concerns.

# the actual check is the one used already in {{globPathIfNecessary}}; that's used in {{DataSource}} and {{FileStreamSource}} and is pretty much the main way wildcards are used to locate files. If it was broken, things would have surfaced.
# its essentially an optimisation to bypass a bit of code which, if you looked at its internals, isn't very efficient. The check for expansion characters is pessimistic, to avoid picking up on wildcard patterns.
# ...but, there are no standalone tests for {{globPathIfNecessary}} —so it may be that there are some failure modes that haven't yet surfaced; regressions waiting to happen.

Given it's a less significant optimisation than dodging all the getFileStatus calls, how about I split that into its own patch, something which should also add those checks for {{globPathIfNecessary}}? I'll trim this patch down to the changes in the streaming package —production code and associated tests

steveloughran committed:
[SPARK-17159] method -> globToFileStatusIfNecessary

SparkQA commented:
**[Test build #64534 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/64534/consoleFull)** for PR 14731 at commit [`4134620`](https://github.com/apache/spark/commit/4134620210e28a2e182397a9bc94ccb8c4d5ffc4).

SparkQA commented:
**[Test build #64534 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/64534/consoleFull)** for PR 14731 at commit [`4134620`](https://github.com/apache/spark/commit/4134620210e28a2e182397a9bc94ccb8c4d5ffc4).
 * This patch **fails PySpark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/64534/
Test FAILed.

steveloughran committed:
[SPARK-17159] File input dstream: revert to directory list operation which doesn't shortcut on a non-wildcard operation

SparkQA commented:
**[Test build #64662 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/64662/consoleFull)** for PR 14731 at commit [`b60f175`](https://github.com/apache/spark/commit/b60f175b5ef058ed24b3ddaf9a85b899a5e33187).

SparkQA commented:
**[Test build #64662 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/64662/consoleFull)** for PR 14731 at commit [`b60f175`](https://github.com/apache/spark/commit/b60f175b5ef058ed24b3ddaf9a85b899a5e33187).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/64662/
Test PASSed.

steveloughran commented:
The latest patch pulls out the shortcutting of the globStatus call if there's no wildcard chars in the path; closer to the original patch

srowen commented on docs/streaming-programming-guide.md:
```diff
@@ -644,13 +644,44 @@ methods for creating DStreams from files as input sources.
+
+     + A simple directory can be supplied, such as `hdfs://namenode:8040/logs/`.
+       All files directly such a path will be processed as they are discovered.
+     + A regular expression can be supplied instead, such as
```
It's a Hadoop-style glob rather than regular expression right? Just checking that "*" in glob would mean "1 or more characters"?

srowen replied:
Also if I read the code correctly, it always treats the input as a glob. If so I think that's how it should be documented. That is, if you give it a path to a file containing "?" it's not going to do what you think. The glob isn't optional if I read it right.

But yeah this looks good to me code-wise.

steveloughran replied:
yes, the input is always a glob. You'd need to \ escape any file with ? in it.

srowen commented on docs/streaming-programming-guide.md:
```diff
@@ -644,13 +644,44 @@ methods for creating DStreams from files as input sources.
+     + The more files under a directory/wildcard pattern, the longer it will take to
+       scan for changes —even if no files have actually changed.
+
+    Special points for object stores
```
These may be so, but what of these are actionable to a user/developer? The point about wildcards is relevant, but what can they do about slow file renames?

steveloughran replied:
w.r.t slow renames, I think people just need to be aware that it can be slow and windows need to plan for it. I'm soon to work on faster renames on S3 [HADOOP-13600](https://issues.apache.org/jira/browse/HADOOP-13600), but it will still be a non-atomic, O(largest-blob) operation

steveloughran replied:
For object stores, direct writes to the directory, resulting in a PUT on close(), will guarantee that a file is picked up immediately. Things are actually a bit quirky for HDFS; even file length doesn't get updated reliably during a write-in-progress. I'll add a section there and then ask people who understand HDFS what is really happening

steveloughran committed:
[SPARK-17159] round out the file streaming text with the dirty details of how HDFS doesn't update file length or modtime until close or a block boundary is reached.

SparkQA commented:
**[Test build #65498 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/65498/consoleFull)** for PR 14731 at commit [`735fc7c`](https://github.com/apache/spark/commit/735fc7c2343c08a323e3d213e611830e3b41ef04).

SparkQA commented:
**[Test build #65498 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/65498/consoleFull)** for PR 14731 at commit [`735fc7c`](https://github.com/apache/spark/commit/735fc7c2343c08a323e3d213e611830e3b41ef04).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/65498/
Test PASSed.

steveloughran committed:
SPARK-17159 Chris Nauroth of HDFS team clarified which operations update the mtime field; this is covered in the streaming section to emphasise why write + rename is the strategy for streaming in files in HDFS. That strategy does also work in object stores, though the rename operation is O(data)

SparkQA commented:
**[Test build #65592 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/65592/consoleFull)** for PR 14731 at commit [`57f697d`](https://github.com/apache/spark/commit/57f697dc718e536f512c856b8e6c8239e1133fd5).

SparkQA commented:
**[Test build #65592 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/65592/consoleFull)** for PR 14731 at commit [`57f697d`](https://github.com/apache/spark/commit/57f697dc718e536f512c856b8e6c8239e1133fd5).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/65592/
Test PASSed.

steveloughran committed:
[SPARK-17159] rebase to master; verify new test still works; review & tighten documentation

SparkQA commented:
**[Test build #66656 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/66656/consoleFull)** for PR 14731 at commit [`f8ed8a3`](https://github.com/apache/spark/commit/f8ed8a3551d1eed5db5a22f5eeb484614036fefe).

SparkQA commented:
**[Test build #66656 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/66656/consoleFull)** for PR 14731 at commit [`f8ed8a3`](https://github.com/apache/spark/commit/f8ed8a3551d1eed5db5a22f5eeb484614036fefe).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/66656/
Test PASSed.

steveloughran commented:
@srowen have you got any comments on the last patch?

srowen commented on docs/streaming-programming-guide.md:
```diff
@@ -644,17 +644,90 @@ methods for creating DStreams from files as input sources.
+     + A POSIX glob pattern can be supplied, such as
+       `hdfs://namenode:8040/logs/2016-??-31`.
+       Here, the DStream will consist of all files directly under those directories
+       matching the regular expression.
```
I think this comment is still incorrect. A glob is not a regex. If this is just the syntax other Hadoop APIs support, that seems reasonable, but it should be described that way.

steveloughran replied:
I added a link to the posix docs. If you follow them, you eventually end up on some coverage of regexps inside []; the Hadoop Glob code does actually convert the shell expression to a java regexp, then compile it in, so presumably should handle everything that the regexp engine (originally {{java.util.regexp}}, currently {{com.google.re2j}} can compile. That's too much detail and something that should really be covered in the Hadoop docs by someone.

srowen commented on docs/streaming-programming-guide.md:
```diff
@@ -644,17 +644,90 @@ methods for creating DStreams from files as input sources.
 	<span class="badge" style="background-color: grey">Python API</span> `fileStream` is not available in the Python API, only	`textFileStream` is	available.
 
+
+    Special points for HDFS
```
Does this need to start a new section in the markdown?

steveloughran replied:
done

srowen commented on docs/streaming-programming-guide.md:
```diff
@@ -644,17 +644,90 @@ methods for creating DStreams from files as input sources.
+    should use PUT operations to directly publish data for a DStream to pick up —even though this
+    is not the mechanism to use when writing to HDFS or other filesystem. When using the Hadoop
+    FileSystem API in Spark that means: write the data directory directly to the target directory,
+    knowing that it is the final `close()` call will make the file visible and set its creation &
```
This is a lot of prose about HDFS semantics. Some of this seems like it's not specific to Spark's usage, and at best should be pointed to from these docs rather than repeated. What about these semantics are necessary to use this API?

steveloughran replied:
I've pulled the details on timestamps into the hadoop docs, and cut back to something that says "filesystems: write then rename, object stores: write directly". That is all people really need to know

srowen reviewed (changes requested):
Am I right that the net change here is not an optimization but an expansion of the behavior to support globs rather than single dirs?

steveloughran committed:
[SPARK-17159] review/tighten docs, move HDFS timestamp details out

SparkQA commented:
**[Test build #70819 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/70819/testReport)** for PR 14731 at commit [`a9a6f7b`](https://github.com/apache/spark/commit/a9a6f7b9e3876e551a2568b6220559992db40228).

SparkQA commented:
**[Test build #70819 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/70819/testReport)** for PR 14731 at commit [`a9a6f7b`](https://github.com/apache/spark/commit/a9a6f7b9e3876e551a2568b6220559992db40228).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/70819/
Test PASSed.

steveloughran commented:
Sean, I think I've managed to delete the lines where you were asking about globs

> Am I right that the net change here is not an optimization but an expansion of the behavior to support globs rather than single dirs?

There's no changes in this source to change the expansion policy; that went in with [SPARK-14976](https://issues.apache.org/jira/browse/SPARK-14976), "make StreamingContext.textFileStream support wildcard". This updates the docs to cover what goes on (the wildcard covering directories, but not the files inside them), and makes the scan much more efficient on object stores. No changes in the semantics of what gets found or when things get found

srowen commented on docs/streaming-programming-guide.md:
```diff
@@ -630,35 +630,106 @@ which creates a DStream from text
+   + Once processed, changes to a file within the current window will not cause the file to be reread.
+     That is: *updates are ignored*.
+   + The more files under a directory, the longer it will take to
+     scan for changes —even if no files have been modified.
```
Nits now: space after hyphen if you please

steveloughran replied:
fixed

srowen commented on docs/streaming-programming-guide.md:
```diff
@@ -630,35 +630,106 @@ which creates a DStream from text
+"Full" Filesystems such as HDFS tend to set the modification time on their files as soon
+as the output stream is created.
+When a file is opened, even before data has been completely written,
+it may be included in the `DStream` &mdash;after which updates to the file within the same window
```
just write a hyphen here and space afterwards

srowen commented on docs/streaming-programming-guide.md:
```diff
@@ -630,35 +630,106 @@ which creates a DStream from text
+{% highlight scala %}
+streamingContext.fileStream[KeyClass, ValueClass, InputFormatClass](dataDirectory)
+{% endhighlight %}
+For text files
```
Space before, and colon at the end of aline?

srowen commented on docs/streaming-programming-guide.md:
```diff
@@ -630,35 +630,106 @@ which creates a DStream from text
+     renaming an entire directory to match the path will add the directory to the list of
+     monitored directories. Only the files in the directory whose modification time is
+     within the current window will be included in the stream.
+   + Calling `FileSystem.setTimes()` to fix the timestamp is a way to have the file picked
```
It could be useful to link to the Hadoop javadoc for this method or at least make clear that this is an HDFS API call, not Spark

srowen commented on docs/streaming-programming-guide.md:
```diff
@@ -630,35 +630,106 @@ which creates a DStream from text
+##### Streaming to FileSystems vs Object stores
+{:.no_toc}
+
+"Full" Filesystems such as HDFS tend to set the modification time on their files as soon
```
Is the message of this section: 

Write new data to a file outside the monitored directory, and then move it into that directory after it's closed. Otherwise, if writing directly to a file in the monitored directory, it may be processed while it's being written to. However: if you are working with an object store like S3, this approach still works correctly but may be slow because a rename can be a copy. In some cases it's OK to write directly into the destination directory because the object store won't expose the file until it's done.

If so I think this could be a little more direct, telling users what they need to know to use this correctly. It seems like renaming is always at least correct. I think that if different advice applies for S3, that should also be called out by name, or else I'm left wondering unnecessarily whether this applies to the most commonly-used store, S3.

steveloughran replied:
FWIW, I've just pulled in some people who have implemented HDFS and S3 clones to give their view of the world, [HADOOP-13946](https://issues.apache.org/jira/browse/HADOOP-13946)

I think for object stores I should actually be vaguer and say "try and it and see". It looks like put ops get their TS when the PUT begins, so multipart put operations will get a TS somewhere between create() and close(). Also, @ajfabbri had to change the semantics on create() to a TS generated by the client when you are using s3guard to cache data.

One troublespot here is that rename() is likely to change timestamps of files in a blobstore, meaning the files window will change from that of creation to that of the rename. That probably is something to call out, as people may need to know of it.

srowen commented on streaming/src/test/scala/org/apache/spark/streaming/InputStreamsSuite.scala:
```diff
@@ -235,18 +236,97 @@ class InputStreamsSuite extends TestSuiteBase with BeforeAndAfter {
+   * Uses the Hadoop APIs to verify consistent behavior with the operations used internally.
+   */
+  test("renamed directories are scanned") {
+    val testDir = Utils.createTempDir()
```
Can you use `withTempDir { testDir =>`? if so I think several tests could use the same pattern here.

steveloughran replied:
That's in `SQLTestUtils`, and not in the streaming suite can pickup without adding spark-sql:test-jar to the test cp. Or: copy & paste in streaming (`org.apache.spark.streaming.TestSuiteBase`), copy & paste somwhere into spark-core:test-jar, which would be a bigger change.

How about I add it to the spark streaming testbase and then move the other tests in InputStreamSuite to it, either in this patch or a successor? That way: cleaner tests

srowen commented on streaming/src/main/scala/org/apache/spark/streaming/dstream/FileInputDStream.scala:
```diff
@@ -196,29 +191,29 @@ class FileInputDStream[K, V, F <: NewInputFormat[K, V]](
-        override def accept(path: Path): Boolean = fs.getFileStatus(path).isDirectory
-      }
-      val directories = fs.globStatus(directoryPath, directoryFilter).map(_.getPath)
+      val directories = Option(fs.globStatus(directoryPath)).getOrElse(Array.empty[FileStatus])
```
Do you need to check for `directories.isEmpty` after this to catch this error case earlier?

steveloughran replied:
I believe the getOrElse clause should handle the globStatus() return value being null, upconverting it to an empty array. Which is essentially working backwards from this special (awful) case in Glob code
```java
    if ((!sawWildcard) && results.isEmpty() &&
        (flattenedPatterns.size() <= 1)) {
      return null;
    }
```
With the directories now being an empty array, `flatMap()` should effectively start with a `map` of the 0-element array, which is will result in the`newFiles` value also being empty. Isn't that right?

srowen reviewed (changes requested):
Mostly some small questions/comments now

uncleGen commented:
@steveloughran Are you still working on this?

steveloughran commented:
let me do a quick review & update

steveloughran committed:
[SPARK-17159] ; address comments, move to withTempDir for tests with a temp dur. Docs now refer reader to the Hadoop FS spec for any details about what object stores do

SparkQA commented:
**[Test build #71866 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/71866/testReport)** for PR 14731 at commit [`06b2bee`](https://github.com/apache/spark/commit/06b2beec75084db1ee330fa4ff4d50775d9f540c).

steveloughran commented:
@uncleGen I've updated it. Note that [HADOOP-13946](https://issues.apache.org/jira/browse/HADOOP-13946) tracks the changes in the Hadoop docs, which writes down what HDFS actually does, then note how cloud object stores have no consistent behaviour w.r.t. timestamps.While I personally believe that direct PUT calls is the way to write data, there's still ambiguity then as to when the objects get a timestamp (S3 : when the PUT/multipart put is first initiated, and not updated on the close() if the put was started earlier) —so when they become visible. So: I don't go into the details, just say "look at the docs, then test on your system". That's about as authoritative as you can get

SparkQA commented:
**[Test build #71866 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/71866/testReport)** for PR 14731 at commit [`06b2bee`](https://github.com/apache/spark/commit/06b2beec75084db1ee330fa4ff4d50775d9f540c).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/71866/
Test PASSed.

steveloughran committed:
SPARK-17159 tail end of the rebase

steveloughran committed:
SPARK-17159 review of docs: quote paths to clearly show what is code and what is just punctuation

steveloughran commented:
@uncleGen: reviewed this, tweaked the docs slightly but otherwise, there's nothing left to do that I can see

SparkQA commented:
**[Test build #73433 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/73433/testReport)** for PR 14731 at commit [`04f4967`](https://github.com/apache/spark/commit/04f49679b3f4f3e2d99e7cafeb9e4fa91fe98ece).

steveloughran committed:
SPARK-17159 address sean's review comments, and read over the object store text and update slightly to make things a bit clearer. The more I learn about object stores, the less they resemble file systems.

SparkQA commented:
**[Test build #73434 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/73434/testReport)** for PR 14731 at commit [`724495b`](https://github.com/apache/spark/commit/724495b97c1521ae5bd4c284d911c5ae6f51b19c).

SparkQA commented:
**[Test build #73433 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/73433/testReport)** for PR 14731 at commit [`04f4967`](https://github.com/apache/spark/commit/04f49679b3f4f3e2d99e7cafeb9e4fa91fe98ece).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/73433/
Test PASSed.

SparkQA commented:
**[Test build #73434 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/73434/testReport)** for PR 14731 at commit [`724495b`](https://github.com/apache/spark/commit/724495b97c1521ae5bd4c284d911c5ae6f51b19c).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/73434/
Test PASSed.

uncleGen commented on docs/streaming-programming-guide.md:
```diff
@@ -615,35 +615,114 @@ which creates a DStream from text
+   + Calling [`FileSystem.setTimes()`](https://hadoop.apache.org/docs/current/api/org/apache/hadoop/fs/FileSystem.html#setTimes-org.apache.hadoop.fs.Path-long-long-)
+     to fix the timestamp is a way to have the file picked up in a later window, even if its contents have not changed.
+
+
```
I remember there is case users may be confused with: if old files are copied from other directories, these files maybe can not be found (as mod_time is not changed). But users are expect to find them. Maybe, we need to doc here?

steveloughran replied:
for "real" filesystems, rename doesn't change modtime, and files become visible in create(), so if you do a create() in the dest dir the file may be found and scanned before the data is fully written. Hence best practise: write elsewhere and rename in. Which by its very nature is the wrong thing to do for object stores: not only is there the perf hit of the copy, the timestamp changes. -and you can't set the times; if you could, we'd probably do it in the filesystem code. PUT-in-place eliminates the copy, and as the file's aren't visible until close(), no risk of early scan

Of course, real (tm) filesystems do let you update the modtime, you can update them after the rename. Even better: do it before so there is no time period when they are in the dest folder with a different timestamp

If you want to make sure this section is clear about what the best practises and surprises are for both FS and blobstore submission, suggest any changes

uncleGen reviewed (approved):
LGTM,  @srowen  give a final OK

steveloughran commented on streaming/src/main/scala/org/apache/spark/streaming/dstream/FileInputDStream.scala:
```diff
@@ -140,7 +137,7 @@ class FileInputDStream[K, V, F <: NewInputFormat[K, V]](
    * modification time returned by the FileStatus API seems to return times only at the
-   * granularity of seconds. And new files may have the same modification time as the
+   * granularity of seconds in HDFS. And new files may have the same modification time as the
    * latest modification time in the previous call to this method yet was not reported in
```
no, it really is HDFS alone. We have no idea whatsoever about the granuarity of other filesystems. Could be 2 seconds (is FAT 32 supported? Hope not) NTFS is in nanoseconds [apparently](https://msdn.microsoft.com/en-us/library/windows/desktop/ms724290(v=vs.85).aspx).

uncleGen replied:
got it

uncleGen commented:
@srowen Waiting for your final OK

steveloughran commented:
The Hadoop FS Spec has now been updated to declare exactly what HDFS does w.r.t timestamps, and warn that what other filesystems and object stores do are implementation and installation specific features: [filesystem.md](https://github.com/apache/hadoop/blob/trunk/hadoop-common-project/hadoop-common/src/site/markdown/filesystem/introduction.md)

That is the associated documentation update with this one; some of the content there was originally here, but moved over to the hadoop docs for the HDFS team to take the blame for when it changes.

steveloughran commented:
Any more comments?

srowen commented on streaming/src/test/scala/org/apache/spark/streaming/TestSuiteBase.scala:
```diff
@@ -557,4 +557,16 @@ trait TestSuiteBase extends SparkFunSuite with BeforeAndAfter with Logging {
+   * (originally from `SqlTestUtils`.)
+   * @todo Probably this method should be moved to a more general place
+   */
+  protected def withTempDir(f: File => Unit): Unit = {
```
We don't already have this defined and available elsewhere?

steveloughran replied:
yes, but in a module that isn't the one where these tests were, so it'd need more dependency logic or pulling it up into a common module, which, if done properly, makes for a big diff

srowen replied:
I see, it's only otherwise defined in the SQL test utils class. Well something we could unify one day, maybe not such a big deal now here.

srowen commented on streaming/src/test/scala/org/apache/spark/streaming/InputStreamsSuite.scala:
```diff
@@ -27,7 +27,8 @@ import scala.collection.JavaConverters._
 
 import com.google.common.io.Files
-import org.apache.hadoop.fs.Path
+import org.apache.commons.io.IOUtils
```
Just use Files.write?

steveloughran replied:
I'll look @ that; I think I went with IOU as it was on the CP and I'd never had bad experiences of it. Guava, well...

steveloughran replied:
Actually, there's a straightforward reason: the test is using the hadoop FS APIs, opening an input stream from a Path and writing to it; `Files.write` is working with a local file. It doesn't work with hadoop FileSystem and Path classes, so could only be used by abusing knowledge of path URLs. Going through FileSystem/Path uses the same API as you'd use in production, so is the more rigorous test.

srowen reviewed (commented):
I have to start from scratch every time I review this ... so this looks like it does more than just optimize a check for new files. It adds docs too. I don't know if the examples are essential. The extra info about how streaming works could be useful but isn't that separate? It's easier to get in small directed changes. This one has been going on for months and I know that's not in anyone's interest.

SparkQA commented:
**[Test build #74990 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/74990/testReport)** for PR 14731 at commit [`a3aaf26`](https://github.com/apache/spark/commit/a3aaf267d2ac30c012b4a71b7a80e28a49ff10be).

SparkQA commented:
**[Test build #74990 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/74990/testReport)** for PR 14731 at commit [`a3aaf26`](https://github.com/apache/spark/commit/a3aaf267d2ac30c012b4a71b7a80e28a49ff10be).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/74990/
Test PASSed.

steveloughran commented:
Is there anything else I need to do here?

steveloughran commented:
@srowen anything else I need to do here?

srowen reviewed (commented):
I've timed out on this change. It changes every time and still doesn't match the title. I don't think this is a great way to pursue changes like this.

steveloughran commented:
Ok. what is the way? Do I write a formal proposal?

Because right now there is no reliable way to get the full dependency graph of Spark + hadoop cloud JARs + direct cloud provider JARs (azure,aws) and their dependencies (jackson) in sync. 

Which means that getting Spark to talk to object stores is more miss than hit.

I'm happy to follow the proposal mechanism, including progress reports &c, but I do at least need some kind of hope that my work will actually get in.

rxin commented:
Steve I think the main point is you should also respect the time of reviewers. The way most of your pull requests manifest have been suboptimal: they often start with a very early WIP (which is not necessarily a problem), and once in a while (e.g. a month or two) you update it to almost completely change it. The time itself is a problem. It requires a lot of context switching to review your pull requests. In addition, every time you update it it looks like a complete new giant pull request.

steveloughran commented:
Reynold, I know very much about the time of reviewers, I put 1+h a day on the hadoop codebase reviewing stuff, generally trying to review the work of non-colleagues, so as to pull in the broad set of contributions which are needed.. 

I have been trying to get some object store related patches into spark alongside the foundational work in fundamentally transforming how we work with object storage, especially S3, in Hadoop. Without the spark side changes, a lot gets lost: here the performance is approx 100-300mS/file when scanning an object store. 

here I've split things in two, docs and diff. Both are independent, both are reasonably tractable. If they can be reviewed fast and added, there's no problems of patches ageing, everyone having to resync.

We can get this out the way, and you've have fewer reasons to be unhappy with me.
